src/Nethermind/Nethermind.JsonRpc/JsonRpcLocalStats.cs

```diff
@@ -16,141 +17,153 @@ namespace Nethermind.JsonRpc
     public class JsonRpcLocalStats : IJsonRpcLocalStats
     {
         private readonly ITimestamper _timestamper;
-        private readonly IJsonRpcConfig _jsonRpcConfig;
         private readonly TimeSpan _reportingInterval;
-
         private ConcurrentDictionary<string, MethodStats> _currentStats = new();
         private ConcurrentDictionary<string, MethodStats> _previousStats = new();
         private readonly ConcurrentDictionary<string, MethodStats> _allTimeStats = new();
         private DateTime _lastReport = DateTime.MinValue;
         private readonly ILogger _logger;
+        private readonly StringBuilder _reportStringBuilder = new();
 
         public JsonRpcLocalStats(ITimestamper timestamper, IJsonRpcConfig jsonRpcConfig, ILogManager logManager)
         {
             _timestamper = timestamper ?? throw new ArgumentNullException(nameof(timestamper));
-            _jsonRpcConfig = jsonRpcConfig ?? throw new ArgumentNullException(nameof(jsonRpcConfig));
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
             _reportingInterval = TimeSpan.FromSeconds(jsonRpcConfig.ReportIntervalSeconds);
         }
 
         public MethodStats GetMethodStats(string methodName) => _allTimeStats.GetValueOrDefault(methodName, new MethodStats());
 
-        public void ReportCall(string method, long handlingTimeMicroseconds, bool success) =>
+        public Task ReportCall(string method, long handlingTimeMicroseconds, bool success) =>
             ReportCall(new RpcReport(method, handlingTimeMicroseconds, success));
 
-        public void ReportCall(in RpcReport report, long elapsedMicroseconds = 0, long? size = null)
+        public Task ReportCall(RpcReport report, long elapsedMicroseconds = 0, long? size = null)
         {
             if (string.IsNullOrWhiteSpace(report.Method))
             {
-                return;
+                return Task.CompletedTask;
             }
 
-            DateTime thisTime = _timestamper.UtcNow;
-            if (thisTime - _lastReport > _reportingInterval)
+            if (!_logger.IsInfo)
             {
-                _lastReport = thisTime;
-                BuildReport();
+                return Task.CompletedTask;
             }
 
-            _currentStats.TryGetValue(report.Method, out MethodStats methodStats);
-            _allTimeStats.TryGetValue(report.Method, out MethodStats allTimeMethodStats);
-            methodStats ??= _currentStats.GetOrAdd(report.Method, _ => new MethodStats());
-            allTimeMethodStats ??= _allTimeStats.GetOrAdd(report.Method, _ => new MethodStats());
+            // we don't want to block RPC calls any longer than required
+            return Task.Run(() =>
```

Comment:
Could just use

```
await Task.Yield();
```

To flip to the threadpool at this point?

Reply:
Don't need to have a lambda sub method in the method; can just use that and make method async rather than Task.Run

Reply:
Ok but why would that be better?

Reply:
agree with Luk here

Reply:
> so introducing async code just to do `Yield` seems odd.

You've already done all that though? So the additional code for the yield is actually less code: https://github.com/NethermindEth/nethermind/commit/0a821a36c15d80f9d8f020044fb9192fe01198bf (delta change) + removing the added indentation from Task.Run

Reply:
> > so introducing async code just to do `Yield` seems odd.
> 
> You've already done all that though? So the additional code for the yield is actually less code: [0a821a3](https://github.com/NethermindEth/nethermind/commit/0a821a36c15d80f9d8f020044fb9192fe01198bf) (delta change) + removing the added indentation from Task.Run

I introduced tasks, but no state machines, no reentrancy.